### PR TITLE
Issue/1226 filename parsing - some path logic.

### PR DESCRIPTION
**Pull request by mitzimorris**
#### Submisison Checklist

- [x] Run tests: `./runCmdStanTests.py src/test`
- [x] Declare copyright holder and open-source license: see below

#### Summary:

Make parsing of filename suffixes more robust - adds logic to handle filepath separators plus some edge cases.

#### Intended Effect:

Correct weirdness reported in #1226 

#### How to Verify:

Unit tests

#### Side Effects:

N/A

#### Documentation:

N/A

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by mitzimorris** [comment]
@WardBrian or @dylex - could you please check on why this failed?

**Comment by WardBrian** [comment]
I think it is because you checked in a change to the `stan` submodule which seems to point to a non-existent commit

File: src/cmdstan/command_helper.hpp [hunk 1]
```diff
@@ -716,6 +780,22 @@ unsigned int get_num_chains(argument_parser &parser) {
 void check_file_config(argument_parser &parser) {
   std::string sample_file
       = get_arg_val<string_argument>(parser, "output", "file");
+  if (sample_file[sample_file.size() - 1] == get_path_separator()
+      || boost::algorithm::ends_with(sample_file, "..")) {
```
Comment (WardBrian) [inline src/cmdstan/command_helper.hpp:784]: I think this could be its own helper `bool is_valid_filename` or similar

File: src/test/interface/command_helper_test.cpp [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+  valid_arguments.push_back(new cmdstan::arg_output());
+  argument_parser parser(valid_arguments);
+  stan::callbacks::stream_writer info(std::cout);
+  stan::callbacks::stream_writer err(std::cerr);
+  int err_code = parser.parse_args(argv.size(), argv_prime, info, err);
+
+  EXPECT_NO_THROW(check_file_config(parser));
+}
+
+TEST(CommandHelper, check_filename_config_fubar) {
```
Comment (WardBrian) [inline src/test/interface/command_helper_test.cpp:88]:
```suggestion
TEST(CommandHelper, check_filename_config_bad) {
```

Perhaps nitpicky, but we use `bad`, `bad2`, etc other places so I think it would be more searchable if these had their names changed

File: src/cmdstan/command_helper.hpp [hunk 3]
```diff
@@ -103,6 +102,64 @@ inline constexpr auto get_arg_val(List &&arg_list, Args &&... args) {
...
+ */
+char get_path_separator() {
+#if defined(WIN32) || defined(_WIN32) \
+    || defined(__WIN32) && !defined(__CYGWIN__)
+  static char path_separator = '\\';
+#else
+  static char path_separator = '/';
+#endif
+  return path_separator;
+}
```
Comment (WardBrian) [inline src/cmdstan/command_helper.hpp:119]: This could be a compile-time constant rather than a function

File: stan
Comment (WardBrian) [inline stan:1]: I believe this is what is breaking CI. Does this branch rely on any changes to the stan submodule?

**Review by WardBrian — CHANGES_REQUESTED** [review]
Thanks for tackling. How difficult do you think it would be to also fix https://github.com/stan-dev/cmdstan/issues/1213?

File: stan
Comment (mitzimorris) [reply stan:1]: no changes to the submodule.

Comment (WardBrian) [reply stan:1]: Ok, I believe restoring the submodule and pushing should unblock CI

**Comment by mitzimorris** [comment]
> How difficult do you think it would be to also fix https://github.com/stan-dev/cmdstan/issues/1213?

not difficult at all - will add to this PR.

File: stan
Comment (mitzimorris) [reply stan:1]: restored?

```
git submodule update --init --recursive --checkout
Submodule path 'stan/lib/stan_math': checked out 'ffeedb5b5d882e5be0c09999089060f16fa8d8e9'
```

Comment (mitzimorris) [reply stan:1]: that didn't work.  wtf?

**Comment by mitzimorris** [comment]
checks are failing because performance tests specify output files with suffix ".tmp". 
submitted PR to cmdstan perf tests repo:  https://github.com/stan-dev/performance-tests-cmdstan/pull/62

@WardBrian - will this work?

File: src/cmdstan/command_helper.hpp [hunk 4]
```diff
@@ -186,11 +259,17 @@ std::vector<std::string> make_filenames(const std::string &filename,
...
                                         unsigned int num_chains,
                                         unsigned int id) {
-  std::vector<std::string> names(num_chains);
-  auto base_sfx = get_basename_suffix(filename);
-  if (base_sfx.second.empty()) {
+  std::pair<std::string, std::string> base_sfx;
+  std::string fname = filename;
+  std::transform(fname.begin(), fname.end(), fname.begin(), ::tolower);
+  if (boost::algorithm::ends_with(fname, ".csv")
+      || boost::algorithm::ends_with(fname, ".json")) {
```
Comment (WardBrian) [inline src/cmdstan/command_helper.hpp:266]: I'm not sure I understand why these are hard coded?

This does seem to fix the example provided in #1213, at least such that output files are properly generated (the `.baz` portion vanishes into thin air, but I suppose this is fine?), there is definitely still some oddities around though. For example,

```
./bernoulli pathfinder save_single_paths=1 output file=foo.bar.baz. data file=bernoulli.data.json
``` 
(note the period at the end of the output file) generates the same files as 
```
./bernoulli pathfinder save_single_paths=1 output file=foo.bar.baz data file=bernoulli.data.json
```

I think some more tests of specifically this function would be helpful!

**Comment by bob-carpenter** [comment]
Could someone please point me to or write doc for how this is supposed to behave?  This PR just says "Make parsing of filename suffixes more robust" and the issue #1226 just lists a bunch of quirky behavior without a spec for what should happen in general.

The problem starts with the name of the argument, `output file=foo/.bar/baz`, because we use this argument to produce an output *directory* and output *base file name prefix*.  If we just took those two things as arguments, we wouldn't need any string processing at all and it'd be clear what's going on to the user.

To allow the current behavior, I would suggest (a) splitting directory off, and (b) operating on the file to remove the final period and anything following it.

The code needs to check that the string-based path is reasonable.  That is, we can't just have `foo/.` as input as that leaves no base file.  We can't have * in the path, etc.  Something needs to validate these are truly paths.

If it didn't break backward compatibility, I would say just slap another `.csv` onto the end of `foo/bar.csv` and be done with it.

**Comment by mitzimorris** [comment]
made all suggested changes, added fix for #1213  - ready for re-review.

**Comment by mitzimorris** [comment]
> Could someone please point me to or write doc for how this is supposed to behave? This PR just says "Make parsing of filename suffixes more robust" and the issue https://github.com/stan-dev/cmdstan/issues/1226 just lists a bunch of quirky behavior without a spec for what should happen in general.

- CmdStan output filename behavoirs
       -  if running a single chain,  the CSV output filename is whatever the `file` argument specifies;  however, if the filename doesn't have suffix `csv` this will be added.
       - if running multiple chains, then the chain id is appended to the end of the filename, e.g. `output_1.csv`
       - for Pathfinder, there's a "save-single-paths" option, this creates a set of per-chain CSV samples which include both tag "_path" as well as chain id - e.g., PSIS sample:  `output.csv`,  plus per-chain `output_path_1.csv`, etc.
       - for Pathfinder, the diagnostic file is JSON format, for the sampler the diagnostic file is CSV format.
     
The CmdStan `output` argument allows 2 sub-arguments:
- `file` - this is the CSV file of draws from the sampler
-  `diagnostic file` - added specifically for the sampler diagnostics and gradients in CSV format

This design (sic) is problematic because:

- multi-chain single-process runs need to use the specified output file names as templates and hack in chain id information
- it was never stipulated that output filenames should end in ".csv"
- Pathfinder can output both the PSIS sample and the inidividual pathfinder samples.
- the diagnostic file from Pathfinder is JSON, not CSV.

W/R/T creating good output filenames:

- this PR fixes the problem that directory names which contain '.' were being parsed as suffixes.
- this PR tries to address problems of specified filenames that are actually just a filepath (per reported quirkiness)
- output file suffixes should be either ".csv" or ".json", according to the output file type.


>The problem starts with the name of the argument, output file=foo/.bar/baz, because we use this argument to produce an output directory and output base file name prefix. If we just took those two things as arguments, we wouldn't need any string processing at all and it'd be clear what's going on to the user.

absolutely.   CmdStanPy only has argument `output_dir` and it handles the logic of naming all the different kinds of output files with unique and consistent names.

>To allow the current behavior, I would suggest (a) splitting directory off, and (b) operating on the file to remove the final period and anything following it.

That's pretty much the way that things are re-imlemented here.  The problem is that C++14 doesn't have good filepath parsing routines and the Boost `filesystem` library is not header-only.  However, C++17 will have a  `filesystem` library, so eventually this will be easier to support.


>The code needs to check that the string-based path is reasonable. That is, we can't just have foo/. as input as that leaves no base file. We can't have * in the path, etc. Something needs to validate these are truly paths.

Agreed.

>If it didn't break backward compatibility, I would say just slap another .csv onto the end of foo/bar.csv and be done with it.

Backwards compatibility is how we got here.  We just fixed an annoying feature of the downstream performance tests which created output files that ended in ".tmp" not ".csv".

File: src/cmdstan/command_helper.hpp (hunk 4 above)
Comment (mitzimorris) [reply src/cmdstan/command_helper.hpp:278]: more unit tests have been added.
more discussion below.

Comment (mitzimorris) [reply src/cmdstan/command_helper.hpp:278]: I don't think that ".baz" will disappear into the air - see line 95 or `command_helper_test.cpp`

**Comment by mitzimorris** [comment]
current behavoir:

```
/examples/bernoulli/bernoulli sample output file=foo.csv. data file=examples/bernoulli/bernoulli.data.json 
```
results in output file `foo.csv..csv`

and 

```
./examples/bernoulli/bernoulli sample output file=foo.bar. data file=examples/bernoulli/bernoulli.data.json
```
results in output file `foo.bar..csv`

**Comment by WardBrian** [comment]
> If we just took those two things as arguments, we wouldn't need any string processing at all and it'd be clear what's going on to the user.

I think that even this would have some subtle issues, like if the 'base file prefix' also contained a path separator. 

I think the fundamental flaw is we should never have set it up like this in the first place. IMO, when multiple files are required, we should have required the user to give us a comma separated list of the correct length and removed any guesswork about "if I specify X as the argument, the actual files created on disk will have the name..."


> To allow the current behavior, I would suggest (a) splitting directory off, and (b) operating on the file to remove the final period and anything following it.

I think this is reasonable. We are also within our rights to say "besides for CSV outputs, we will force the correct extension". So the bug reported in #1213 would yield `foo.bar_path_1.baz` as a CSV file, but `foo.bar_path_1.json` as the diagnostic file. That would maintain backward compatibility (by allowing csv outputs to still have arbitrary extensions) but be a bit more sane going forward with new formats

File: src/cmdstan/command_helper.hpp (hunk 4 above)
Comment (WardBrian) [reply src/cmdstan/command_helper.hpp:278]: After pulling, and running `./bernoulli pathfinder save_single_paths=1 output file=foo.bar.baz data file=bernoulli.data.json`:

```
$ ls
bernoulli            bernoulli.data.R  foo.bar.csv         foo.bar_path_1.json  foo.bar_path_2.json  foo.bar_path_3.json  foo.bar_path_4.json
bernoulli.data.json  bernoulli.stan    foo.bar_path_1.csv  foo.bar_path_2.csv   foo.bar_path_3.csv   foo.bar_path_4.csv
```

no `baz` to be found! This is concerning, because I agree the test suggests otherwise!

Comment (WardBrian) [reply src/cmdstan/command_helper.hpp:278]: This is caused by the `output_base` variable in `command.hpp`. Integration test may be needed, sadly

File: src/cmdstan/command.hpp [hunk 5]
```diff
@@ -197,34 +197,33 @@ int command(int argc, const char *argv[]) {
...
         && (get_arg_val<bool_argument>(parser, "method", "pathfinder",
                                        "save_single_paths"));
-  bool save_diagnostics = diagnostic_base != output_base;
 
   if (user_method->arg("pathfinder")) {
+    if (save_single_paths && diagnostic_file.empty()) {
+      diagnostic_file = output_file;
+    }
+    std::cout << "save_single_paths ? " << save_single_paths
+              << " diagnostic file? " << diagnostic_file << std::endl;
```
Comment (WardBrian) [inline src/cmdstan/command.hpp:210]:
```suggestion
```

File: src/cmdstan/command.hpp [hunk 6]
```diff
@@ -197,34 +197,33 @@ int command(int argc, const char *argv[]) {
...
         init_filestream_writers(diagnostic_json_writers, num_chains, id,
-                                diagnostic_base, "", ".json", sig_figs);
+                                diagnostic_file, "", ".json", sig_figs);
       } else {
         init_null_writers(diagnostic_json_writers, num_chains);
       }
     } else {
-      if (save_single_paths || save_diagnostics) {
-        init_filestream_writers(sample_writers, num_chains, id, output_base,
+      if (save_single_paths) {
```
Comment (WardBrian) [inline src/cmdstan/command.hpp:222]: To get exactly the same behavior as before with these arguments, you need to still have an or condition here

```suggestion
      if (save_single_paths || !diagnostic_file.empty()) {
```

File: src/cmdstan/command.hpp [hunk 7]
```diff
@@ -338,7 +336,7 @@ int command(int argc, const char *argv[]) {
           save_single_paths, refresh, interrupt, logger, init_writer,
           sample_writers[0], diagnostic_json_writers[0]);
     } else {
-      auto ofs = std::make_unique<std::ofstream>(output_base + ".csv");
+      auto ofs = std::make_unique<std::ofstream>(output_file);
```
Comment (WardBrian) [inline src/cmdstan/command.hpp:339]: using make_filenames and taking the first element should make the behavior more consistent with other methods

File: src/test/interface/pathfinder_test.cpp [hunk 8]
```diff
@@ -18,7 +18,7 @@ class CmdStan : public testing::Test {
     eight_schools_data
         = {"src", "test", "test-models", "eight_schools.data.json"};
     empty_model = {"src", "test", "test-models", "empty"};
-    arg_output = {"test", "output"};
+    arg_output = {"test", "output.csv"};
```
Comment (WardBrian) [inline src/test/interface/pathfinder_test.cpp:21]: Revert after making above change

**Review by WardBrian — CHANGES_REQUESTED** [review]
A few small things, should be last review. Thanks!

**Review by WardBrian — APPROVED** [review]
Thanks again for tackling!
